File: clojure/src/project_euler/sol/p0024.clj
(ns project-euler.sol.p0024
  (:require
   [project-euler.lib.math :as my-math]))

(defn solve
  ([]
   (solve 1000000))
  ([ith]
   (loop [nodes (vec (range 10))
          idx (dec ith)
          acc 0]
     (if-not (empty? nodes)
       (let [blk (my-math/factorial (dec (count nodes)))
             cnt (quot idx blk)
             n (nth nodes cnt)]
         (recur (vec (remove #(= % n) nodes))
                (mod idx blk)
                (+ (* acc 10) n)))
       (format "%010d" acc)))))

File: haskell/src/Sol/P0024.hs
module Sol.P0024 (compute, solve) where

import Text.Printf (printf)

import Mylib.Math (factorial)

compute :: Int -> String
compute nth =
    printf "%010d" $ aux [0 .. 9] (pred nth) 0
  where
    aux :: [Int] -> Int -> Int -> Int
    aux [] _ acc = acc
    aux xs idx acc =
        aux (hd ++ (drop 1 tl)) (idx `mod` blk) (10 * acc + xs !! cnt)
      where
        blk = factorial (pred $ length xs)
        cnt = idx `div` blk
        (hd, tl) = splitAt cnt xs

solve :: String
solve = compute 1_000_000
